pydeseq2/dds.py

```diff
@@ -593,22 +569,39 @@ def calculate_cooks(self):
         """
 
         # Check that MAP dispersions are available. If not, compute them.
-        if not hasattr(self, "dispersions"):
+        if "dispersions" not in self.varm:
             self.fit_MAP_dispersions()
 
-        num_vars = self.design_matrix.shape[1]
+        num_vars = self.obsm["design_matrix"].shape[-1]
+        nz_data = self[:, self.non_zero_genes]
+
         # Keep only non-zero genes
-        normed_counts = self.counts.loc[:, self.non_zero_genes].div(self.size_factors, 0)
-        dispersions = robust_method_of_moments_disp(normed_counts, self.design_matrix)
-        V = self._mu_LFC + dispersions * self._mu_LFC**2
-        squared_pearson_res = (
-            self.counts.loc[:, self.non_zero_genes] - self._mu_LFC
-        ) ** 2 / V
-        self.cooks = (
+        normed_counts = pd.DataFrame(
+            nz_data.X / self.obsm["size_factors"][:, None],
+            index=self.obs_names,
+            columns=self.non_zero_genes,
+        )
+
+        # dispersions = pd.Series(np.NaN, index=self.var_names)
+        dispersions = robust_method_of_moments_disp(
+            normed_counts, self.obsm["design_matrix"]
+        )
+
+        V = (
+            nz_data.layers["_mu_LFC"]
+            + dispersions.values[None, :] * nz_data.layers["_mu_LFC"] ** 2
+        )
+        squared_pearson_res = (nz_data.X - nz_data.layers["_mu_LFC"]) ** 2 / V
+
+        self.layers["cooks"] = np.full((self.n_obs, self.n_vars), np.NaN)
+        self.layers["cooks"][:, self.non_zero] = (
             squared_pearson_res
             / num_vars
-            * (self._hat_diagonals / (1 - self._hat_diagonals) ** 2)
-        ).T
+            * (
+                nz_data.layers["_hat_diagonals"]
+                / (1 - nz_data.layers["_hat_diagonals"]) ** 2
+            )
+        )
```

Comment:
with the linting, the expression becomes a bit hard to follow. What about cutting it in pieces?

Reply:
done in [458d9d8](https://github.com/owkin/PyDESeq2/pull/66/commits/458d9d87fc25100e6bf44eefb1d828621cd61bd1)